upcxx: Update the UPC++ package to 2021.9.0

bonachea opened the pull request:
Changes: 

* Add the new release, and a missing older one.

* Remove the spack package cruft for supporting the obsolete build system that
  was present in older versions that are no longer supported.

* General cleanups.

Support for library versions older than 2020.3.0 is officially retired,
for two reasons:

1. Releases prior to 2020.3.0 had a required dependency on Python 2,
   which is [officially EOL](https://www.python.org/doc/sunset-python-2/)
   as of Jan 1 2020, and is no longer considered secure.
2. (Most importantly) The UPC++ development team is unable/unwilling to
   support releases more than two years old.  UPC++ provides robust
   backwards-compatibility to earlier releases of UPC++ v1.0, with very
   rare well-documented/well-motivated exceptions.  Users are strongly
   encouraged to update to a current version of UPC++.

NOTE: Most of the lines changed in this commit are simply re-indentation,
and thus might be best reviewed in a diff that ignores whitespace.

bonachea committed:
upcxx: Update the UPC++ package to 2021.9.0

* Add the new release, and a missing older one.

* Remove the spack package cruft for supporting the obsolete build system that
  was present in older versions that are no longer supported.

* General cleanups.

Support for library versions older than 2020.3.0 is officially retired,
for two reasons:

1. Releases prior to 2020.3.0 had a required dependency on Python 2,
   which is [officially EOL](https://www.python.org/doc/sunset-python-2/)
   as of Jan 1 2020, and is no longer considered secure.
2. (Most importantly) The UPC++ development team is unable/unwilling to
   support releases more than two years old.  UPC++ provides robust
   backwards-compatibility to earlier releases of UPC++ v1.0, with very
   rare well-documented/well-motivated exceptions.  Users are strongly
   encouraged to update to a current version of UPC++.

NOTE: Most of the lines changed in this commit are simply re-indentation,
and thus might be best reviewed in a diff that ignores whitespace.

PHHargrove reviewed (dismissed):
Looks good to me.

bernhardkaindl commented:
@bonachea This PR is currently a draft and we can't merge it as draft. When you are ready for merge/maintainer-review click the button labelled "Ready for review" above the list of passed checks right of this text:
> This pull request is still a work in progress
Draft pull requests cannot be merged.

bonachea commented:
> @bonachea This PR is currently a draft and we can't merge it as draft. When you are ready for merge/maintainer-review click the button labelled "Ready for review" above the list of passed checks right of this text:

Thanks @bernhardkaindl - One of our maintainers is on vacation this week, so I've deliberately left this PR in draft mode to keep it "on hold" until he returns, allowing him to be involved in any potential discussion with reviewers.

bonachea committed:
upcxx: Detect Cray XC more explicitly

This change is necessary to prevent false matches occuring on new Cray Shasta
systems, which do not use the aries network but were incorrectly being treated
as a Cray XC + aries platform.

UPC++ has not yet deployed official native support for Cray Shasta, but this
change is sufficient to allow building the portable backends there.

bonachea commented:
I've just rebased onto the latest develop and appended a commit that fixes an incompatibility on new Cray Shasta systems.
